src/qibo/abstractions/states.py

```diff
@@ -0,0 +1,267 @@
+
+
+class AbstractState(ABC):
+    """Abstract class for quantum states returned by model execution.
+
+    Args:
+        nqubits (int): Optional number of qubits in the state.
+            If ``None`` then the number is calculated automatically from the
+            tensor representation of the state.
+    """
+
+    def __init__(self, nqubits=None):
+        self._nqubits = None
+        self._tensor = None
+        self.nstates = None
+        self.measurements = None
+        if nqubits is not None:
+            self.nqubits = nqubits
+
+    @property
+    def nqubits(self):
+        """Number of qubits in the state."""
+        if self._nqubits is None:
+            raise_error(AttributeError, "State number of qubits not available.")
+        return self._nqubits
+
+    def __len__(self):
+        """Number of components in the state's tensor representation."""
+        if self._nqubits is None:
+            raise_error(AttributeError, "State number of qubits not available.")
+        return self.nstates
+
+    @nqubits.setter
+    def nqubits(self, n):
+        self._nqubits = n
+        self.nstates = 2 ** n
+
+    @property
+    @abstractmethod
+    def shape(self): # pragma: no cover
+        """Shape of the state's tensor representation."""
+        raise_error(NotImplementedError)
+
+    @property
+    @abstractmethod
+    def dtype(self): # pragma: no cover
+        """Type of state's tensor representation."""
+        raise_error(NotImplementedError)
+
+    @property
+    def tensor(self):
+        """Tensor representation of the state in the computational basis."""
+        if self._tensor is None:
+            raise_error(AttributeError, "State tensor not available.")
+        return self._tensor
+
+    @tensor.setter
+    def tensor(self, x):
+        nqubits = int(math.log2(len(x)))
+        if self._nqubits is None:
+            self.nqubits = nqubits
+        elif self._nqubits != nqubits:
+            raise_error(ValueError, "Cannot assign tensor of length {} to "
+                                    "state with {} qubits."
+                                    "".format(len(x), self.nqubits))
+        self._tensor = x
+
+    @abstractmethod
+    def __array__(self): # pragma: no cover
+        """State's tensor representation as an array."""
+        raise_error(NotImplementedError)
+
+    @abstractmethod
+    def numpy(self): # pragma: no cover
+        """State's tensor representation as a numpy array."""
+        raise_error(NotImplementedError)
+
+    def state(self, numpy=False):
+        """State's tensor representation as an backend tensor.
+
+        Args:
+            numpy (bool): If ``True`` the returned tensor will be a numpy array,
+                otherwise it will follow the backend tensor type.
+                Default is ``False``.
+        """
+        if numpy:
+            return self.numpy()
+        return self.tensor
+
+    def __repr__(self):
+        return str(self.tensor)
```

Comment:
Ok, after playing with the code I think that we should move that to `.summary()` method, and keep the default `__repr__`, because the user may get confused about the return type of `Circuit.__call__`.